Introduce BPF Streams abstraction

etsal opened the pull request:
Add libbpf-rs support for interacting with BPF streams from Rust code. The interface provides the option to either directly specify the prog fd/stream id, or to hide the details behind a Stream abstraction.

danielocfb commented on libbpf-rs/src/streams.rs:
```diff
@@ -0,0 +1,105 @@
+    /// unnecessarily complicated..
+    fn opts_dfl() -> StreamReadOpts {
+        StreamReadOpts { sz: 0 }
+    }
```
This constructor should probably be replaced by a [`Default`](https://doc.rust-lang.org/std/default/trait.Default.html) impl?

danielocfb commented on libbpf-rs/src/streams.rs:
```diff
@@ -0,0 +1,105 @@
+/// Config options, currently empty and here for forward compatibility.
+struct StreamReadOpts {
+    sz: usize,
+}
```
I kind of fail to see the purpose of this type. It's not public, so users can't use it. So we don't need forward compat. And if it's empty, then we don't need it at all?

danielocfb commented on libbpf-rs/src/streams.rs:
```diff
@@ -0,0 +1,105 @@
+    const BYTES_PER_READ_DFL: usize = 4096;
+
+    /// Create a new Stream instance.
+    pub fn new(fd: i32, stream_id: u32) -> Self {
```
So what is a "stream_id" and what is its purpose? It's unclear what relevance it has to the user. I don't see it mentioned in the only "documentation" that I can find on the topic, which is https://lwn.net/Articles/1026660/ (and its reincarnations)

danielocfb reviewed (changes requested):
Hi @etsal, thanks for the pull request. Left a few comments with suggestions for improvement/clarifying questions. Can you please make sure to add tests for the functionality?

In general it feels as if `Stream` should likely implement [`io::Read`](https://doc.rust-lang.org/std/io/trait.Read.html) to enable proper interop with other Rust components.

danielocfb commented on libbpf-rs/src/streams.rs:
```diff
@@ -0,0 +1,105 @@
+/// interface from the BPF program to userspace.
+#[derive(Debug)]
+pub struct Stream {
+    fd: i32,
```
Rust generally uses file descriptor abstractions that encode ownership semantics see https://doc.rust-lang.org/stable/std/os/fd/index.html

etsal replied:
The fd in question is the fd of the program whose stream we are reading. AFAICT the real fd is owned by the ProgramMut. Would it work to just embed the `Stream` into `ProgramMut`, then turn this fd into a BorrowedFd? I avoided doing so because `ProgramMut` is a pretty thin wrapper around the underlying C structs, but I think that would be the most natural place for `Stream`.

etsal replied:
Actually, combining this with the comment about stdout() above: We don't need to embed anything, we can just add a method for a ProgramMut that creates a Stream object for its stdout or stderr. That would address both comments.

etsal commented:
Thank you for the review, I will update the PR accordingly and post here to confirm it's ready for review.  Ack to all comments, though I have a couple  followup questions (inline).

etsal commented:
Updated the code and added tests. The interface now presents the streams only indirectly and without exposing the type itself unless the user opts to create an instance directly. I think the code now is more idiomatic and looks a lot less like C.

danielocfb commented on libbpf-rs/tests/test_streams.rs:
```diff
@@ -0,0 +1,66 @@
+    let mut stdout = prog.stdout();
+    let mut buf = [0u8; 1024];
+
+    // The read itself should succeed and return 0 bytes
```
Am I misreading or does the comment state the opposite of what is being asserted?

danielocfb commented on libbpf-rs/tests/bin/src/stream.bpf.c:
```diff
@@ -0,0 +1,65 @@
+
+volatile u64 i;
+
+/* Trigger a may_goto timeout to emit a streams error. */
```
So this `may_goto` timeout is a way to force the kernel to emit data to stderr, is that how it works? And there is no way to write to a stream directly? And what are the requirements for `may_goto` to be available? Is it just a certain LLVM version? Or kernel config? It's also not exactly clear what happens if it's not available. Will the test still work?

If there is no easier way to trigger a stream write, can you please expand the comment a bit to make it clear who does what and in response to what exactly?

etsal commented:
Not sure why clippy is hitting these errors, they are for pre-existing code.

etsal commented:
Thank you for the detailed feedback!
